ironfish/src/network/transactionFetcher.ts

```diff
@@ -0,0 +1,202 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+import { BufferMap } from 'buffer-map'
+import { TransactionHash } from '../primitives/transaction'
+import { PooledTransactionsRequest } from './messages/pooledTransactions'
+import { PeerNetwork } from './peerNetwork'
+import { Peer, PeerState } from './peers/peer'
+
+type TransactionState =
+  | {
+      action: 'REQUEST_SCHEDULED'
+      timeout: NodeJS.Timeout
+    }
+  | {
+      peer: Peer
+      action: 'IN_FLIGHT'
+      timeout: NodeJS.Timeout
+      clearDisconnectHandler: () => void
+    }
+  | {
+      action: 'PROCESSING'
+    }
+
+/**
+ * When a node receives a new transaction hash it needs to query the sender for
+ * the full transaction object. This class encapsulates logic for resolving transaction
+ * hashes to full transactions from the network. It operates as a state machine. Each transaction
+ * hash has it's own state which changes when peers are queried, requests timeout or a
+ * full transaction is received from the network
+ */
+export class TransactionFetcher {
+  // State of the current requests for each transaction
+  private readonly pending = new BufferMap<TransactionState>()
+
+  // Set of peers that also may be able to fetch the transaction
+  private readonly sources = new BufferMap<Set<Peer>>()
```

Comment:
I think we should store peerIdentities rather than peers here, to avoid preventing Peers from getting garbage collected.